### PR TITLE
GIF support using FLAnimatedImage

### PR DESCRIPTION
**Pull request by bpoplauschi**
* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [ ] I have added the required tests to prove the fix/feature I am adding
* [ ] I have updated the documentation (if necessary)
* [x] I have run the tests and they pass
* [x] I have run the lint and it passes (`pod lib lint`)

This merge request fixes: #945 and #733 and a bunch of [other GIF related problems](https://github.com/rs/SDWebImage/issues?utf8=%E2%9C%93&q=label%3AGIF+)

### Pull Request Description

Using the FLAnimatedImage (very nice lib from Flipboard), redone our GIF support. Inspired from integrations like: https://gist.github.com/hilen/8ef4a810406051e8f0c1

Notes:
- the integration was not as smooth as I would have wanted, but this is because the way FLAnimatedImage was built
- `FLAnimatedImageView` extends `UIImageView`, but `FLAnimatedImage` does not extend `UIImage`. 
- also, there is some *magic* happening inside `FLAnimatedImageView` that allows this great performance. But that also adds a dependency for people to use this class instead of `UIImageView`. That's why I created the extra category.

The solution is pretty simple:
- created a `FLAnimatedImageView (WebCache)` category that hooks to the `SDWebImageManager`. At this point, I tried (and worked for me) overwriting the methods from the `UIImageView (WebCache)` category in the `FLAnimatedImageView (WebCache)` category. It works fine, but we need to make sure this happens all the time. There seems to be a difference from the case where there are 2 categories for the same class which makes the result random (depends on the categories loading in memory order).
- changed our old GIF support to return static images so they still work that way for backwards compatibility. 
- using the `images` array so we can later distinguish between static images and those static from GIF images. We need to distinguish because we need the NSData in all cases so we can pass it to the `FLAnimatedImage` class. Even in the case where there is an `UIImage` in the memory cache.
- had to make some adjustments to our components to be able to have access to the NSData
- updated the example

### CODE REVIEW

**Comment by bpoplauschi** [comment]
@mythodeia @rs could you check this out and let me know what you think?

File: SDWebImage/SDImageCache.m [hunk 1]
```diff
@@ -409,14 +414,15 @@ - (NSOperation *)queryDiskCacheForKey:(NSString *)key done:(SDWebImageQueryCompl
         }
 
         @autoreleasepool {
+            NSData *diskData = [self diskImageDataBySearchingAllPathsForKey:key];
             UIImage *diskImage = [self diskImageForKey:key];
             if (diskImage && self.shouldCacheImagesInMemory) {
                 NSUInteger cost = SDCacheCostForImage(diskImage);
                 [self.memCache setObject:diskImage forKey:key cost:cost];
             }
```
Comment (bpoplauschi) [inline SDWebImage/SDImageCache.m]: the code above is not optimal, I know. I think I should change it. We explicitly call `diskImageDataBySearchingAllPathsForKey:` that loads from disk and then call `diskImageForKey:` that also calls `diskImageDataBySearchingAllPathsForKey:` and transforms that into a UIImage. Too bad we don't have tuples like in Swift, we could just change the `diskImageDataBySearchingAllPathsForKey:` to return (UIImage *, NSData *)

File: SDWebImage/SDImageCache.m [hunk 2]
```diff
@@ -409,14 +414,15 @@ - (NSOperation *)queryDiskCacheForKey:(NSString *)key done:(SDWebImageQueryCompl
         }
 
         @autoreleasepool {
+            NSData *diskData = [self diskImageDataBySearchingAllPathsForKey:key];
```
Comment (bpoplauschi) [inline SDWebImage/SDImageCache.m]: the code above is not optimal, I know. I think I should change it. We explicitly call `diskImageDataBySearchingAllPathsForKey:` that loads from disk and then call `diskImageForKey:` that also calls `diskImageDataBySearchingAllPathsForKey:` and transforms that into a UIImage. Too bad we don't have tuples like in Swift, we could just change the `diskImageDataBySearchingAllPathsForKey:` to return (UIImage *, NSData *)

**Comment by mythodeia** [comment]
as always great work @bpoplauschi. i will check it out this weekend

**Comment by NSKevin** [comment]
Awesome! By the way, I think you can add a safe check at SDWebImageManager.m line:61 . If the url is nil, there will be a crash here. Thanks~

**Comment by bpoplauschi** [comment]
Merging this to avoid keeping multiple branches. Will make all future changes directly on the 4.x branch. Also need to update the documentation (and the migration guide #1590 )

**Comment by bpoplauschi** [comment]
@mythodeia please still review this when you have time.

**Comment by mythodeia** [comment]
@bpoplauschi 
so far i have tested on a 6Plus & 6S Plus with a dozen gifs and i have no crashes yet.
of course there is an increase in memory  and CPU while the gif cells are showing.if i scroll it goes back to normal percents.
gifs showing on screen on an iPhone 6SPlus:
gifs

gifs showing on screen on an iPhone 6S:
gifs6s

no gifs on screen:
nogifs

**Comment by NSKevin** [comment]
@mythodeia I have a lot of Gif in my UITableView and When I select one, it will turn into big photo browse mode. It still crash with some gif and the log show that：

>Type:  00000020
Exception Codes: 0x000000008badf00d
Exception Note:  SIMULATED (this is NOT a crash)
Highlighted by Thread:  0

>Application Specific Information:
com.babytree.lama failed to launch after 20.00s (launchIntent: foreground-interactive)

>Elapsed total CPU time (seconds): 1.770 (user 1.770, system 0.000), 4% CPU 
Elapsed application CPU time (seconds): 0.000, 0% CPU

>Filtered syslog:
None found

>Thread 0:
0   dyld                          	0x0000000120051000 _dyld_start + 0

>No thread state (register information) available

>Binary Images:
0x120050000 - 0x12007ffff dyld arm64  <b9ab8d2475913f8c9974747a7feab615> /usr/lib/dyld

**Comment by mythodeia** [comment]
@NSKevin can you share any gif that produces this crash?

**Comment by bpoplauschi** [comment]
@mythodeia @NSKevin this crash doesn't look like a regular crash. It's a simulated (not sure what that means) watchdog crash, not related to memory usage.

**Comment by NSKevin** [comment]
Terminated due to Memory Pressure

**Comment by eggswift** [comment]
@bpoplauschi 
Why you change `sd_animatedGIFWithData:` method to return 1frame image?
Maybe need compatibility with UIImageView.

#1836

**Comment by bpoplauschi** [comment]
@eggswift Because the idea was to force people using the `FLAnimatedImage` component for the GIFs, as the one we had before had poor performance.
Hope this makes sense. I will close the ticket #1836 and continue the conversation here, where other people involved can see it.

**Comment by eggswift** [comment]
@bpoplauschi 
Thanks for this help very much : ) 

I know that you recommend people to use `FLAnimatedImage ` for the GIFs, but as I know, FLAnimatedImage still have some todo works, especially support other animated image formats such as APNG or WebP, this requirement hit me so much in our project.

I think there r lots of people using `UIImageView` to display GIFs now, and it seems that `UIImageView ` can not display GIFs in SDWebImage 4.x.x. It's hard for us to upgrade it.

So I think we need to do a compatibility with `UIImageView ` until `FLAnimatedImage` can works well with WebP or APNG and other requirement.

What do you think? Thanks : )

**Pull request merged**